### PR TITLE
feat: core plugin support

### PR DESCRIPTION
**Pull request by dylandepass**
First draft at plugin support.

### Plugins Added

- [x] Environment Switcher
- [x] Edit
- [x] Preview
- [x] Publish

### To do

- [ ] Add tests

### Not covered in this PR
Custom plugins
Advanced plugins (only visible in advanced mode)
I'm sure there is more...

### CODE REVIEW

**Review by rofe — CHANGES_REQUESTED** [review]

File: src/extension/types/typedefs.js [hunk 1]
```diff
@@ -110,6 +110,45 @@
  * @description The plugin configuration.
  */
 
+/**
+ * @typedef {Object} _Plugin
+ * @prop {string} id The plugin ID (mandatory)
+ * @prop {PluginButton} [button] A button configuration object (optional)
+ * @prop {string} [container] The ID of a dropdown to add this plugin to (optional)
+ * @prop {boolean} [feature] Determines whether to group this plugin with the features (default: false)
```
[the remark this answers is not in the document]
Comment (dylandepass) [reply src/extension/types/typedefs.js:118]: From what I can tell from the designs I don't think there is a "feature" container. Besides the plugin-picker (sorter) and profile button, everything looks to just be "plugins". 

I think we can remove

File: src/extension/types/typedefs.js [hunk 2]
```diff
@@ -110,6 +110,45 @@
...
+ * @prop {boolean} [override] Determines whether to replace an existing plugin (default: false)
```
[the remark this answers is not in the document]
Comment (dylandepass) [reply src/extension/types/typedefs.js:119]: Agreed, can you do complete overrides or just override the condition?

File: src/extension/types/typedefs.js [hunk 3]
```diff
@@ -110,6 +110,45 @@
...
+ * @prop {ElemConfig[]} [elements]  An array of elements to add (optional)
```
[the remark this answers is not in the document]
Comment (dylandepass) [reply src/extension/types/typedefs.js:120]: I'm not seeing anyone using the feature in `adobecom` or `hlxsites`. Are you aware of anyone using it? Curious to see how it's being used.

File: src/extension/types/typedefs.js [hunk 4]
```diff
@@ -110,6 +110,45 @@
...
+ * @prop {Function} [condition] Determines whether to show this plugin (optional).
+ * This function is expected to return a boolean when called with the sidekick as argument.
+ * @prop {Function} [advanced] Show this plugin only in advanced mode (optional).
```
[the remark this answers is not in the document]
Comment (dylandepass) [reply src/extension/types/typedefs.js:123]: Sounds right, will remove

File: src/extension/types/typedefs.js [hunk 5]
```diff
@@ -110,6 +110,45 @@
...
+
+/**
+ * @typedef {Object} ElemConfig
+ * @private
+ * @prop {string} tag The tag name (mandatory)
+ * @prop {string} text The text content (optional)
+ * @prop {Object[]} attrs The attributes (optional)
+ * @prop {Object[]} lstnrs The event listeners (optional)
+ * @description The configuration of an element to add.
+ */
```
[the remark this answers is not in the document]
Comment (dylandepass) [reply src/extension/types/typedefs.js:142]: I have brought back `createTag` as a shorthand to creating HTMLElements. So I think we still need it.
https://github.com/adobe/aem-sidekick/pull/44/files#diff-706f9ae15c33668b11a311c20ef407c9dec13b61c7d5a9a1ba2e85d7c1b80589R161

**Comment by dylandepass** [comment]
@rofe In terms of custom events dispatched from plugins.. This maybe tricky to juggle due to the name change of the custom element (helix-sidekick -> aem-sidekick). Guess will have to get people to add event listeners to both elements until `helix-sidekick` is decommisioned.

https://github.com/adobecom/milo/blob/13d29054db249a66b8a257c045d996f7e213c660/libs/utils/sidekick.js#L32

**Review by rofe — APPROVED** [review]

**Pull request merged**